File: crates/storage/src/operations/usage.rs
use std::fmt;

use chrono::{DateTime, Utc};
use olp_domain::OperationKind;
use uuid::Uuid;

mod breakdown;
mod completeness;
pub(super) mod query;
mod series;
mod summary;

pub use breakdown::{UsageBreakdown, UsageBreakdownReport};
pub use completeness::UsageCompleteness;
pub use series::{UsagePoint, UsageSeries};
pub use summary::UsageSummary;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum UsageGranularity {
    Hour,
    Day,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum UsageDimension {
    Route,
    Provider,
    Model,
    ApiKey,
    Operation,
}

#[derive(Clone, Debug)]
pub struct UsageFilters {
    pub observed_after: DateTime<Utc>,
    pub observed_before: DateTime<Utc>,
    pub route_slug: Option<String>,
    pub provider_id: Option<Uuid>,
    pub upstream_model: Option<String>,
    pub api_key_id: Option<Uuid>,
    pub operation: Option<OperationKind>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct UsageRangeCoverage {
    /// False only when a requested partial hour exists solely as a retained
    /// hourly aggregate and therefore cannot be sliced without guessing.
    pub range_complete: bool,
    /// Signals that returned totals cover only the exact, representable subset
    /// of the requested range. OLP never prorates hourly aggregates.
    pub approximate: bool,
    pub excluded_partial_aggregate_boundaries: u8,
}

impl fmt::Display for UsageDimension {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Route => "route",
            Self::Provider => "provider",
            Self::Model => "model",
            Self::ApiKey => "api_key",
            Self::Operation => "operation",
        })
    }
}

File: console/src/lib/api/usage.ts
import type { components } from '$lib/api/schema';
import { apiClient } from '$lib/api/client';
import { result } from '$lib/api/http';
import { compactQuery } from '$lib/api/query';

export type UsagePoint = components['schemas']['UsagePointResponse'];
export type UsageCompleteness =
  components['schemas']['UsageCompletenessResponse'];

export type UsageFilters = {
  start: string;
  end: string;
  route?: string;
  provider_id?: string;
  model?: string;
  api_key_id?: string;
  operation?: string;
};

type UsageSummary = components['schemas']['UsageSummaryResponse'];
type UsageSeriesResult = components['schemas']['UsageTimeSeriesResponse'];
type UsageBreakdownResult = components['schemas']['UsageBreakdownResponse'];

export async function usageSummary(
  filters: UsageFilters
): Promise<UsageSummary> {
  const { data, error, response } = await apiClient.GET(
    '/api/v3/usage/summary',
    {
      params: { query: compactQuery(filters) }
    }
  );
  return result(data, error, response);
}

export async function usageSeries(
  filters: UsageFilters,
  granularity: 'hour' | 'day'
): Promise<UsageSeriesResult> {
  const { data, error, response } = await apiClient.GET(
    '/api/v3/usage/time-series',
    { params: { query: compactQuery({ ...filters, granularity }) } }
  );
  return result(data, error, response);
}

export async function usageBreakdown(
  filters: UsageFilters,
  dimension: 'route' | 'provider' | 'model' | 'api_key' | 'operation'
): Promise<UsageBreakdownResult> {
  const { data, error, response } = await apiClient.GET(
    '/api/v3/usage/breakdown',
    { params: { query: compactQuery({ ...filters, dimension, limit: 50 }) } }
  );
  return result(data, error, response);
}

export async function usageCompleteness(
  filters: UsageFilters
): Promise<UsageCompleteness> {
  const { data, error, response } = await apiClient.GET(
    '/api/v3/usage/completeness',
    { params: { query: compactQuery(filters) } }
  );
  return result(data, error, response);
}
